[PCluster Configure] Add integration tests for enabling EFA through the pcluster configure wizard

EddyMM opened the pull request:
### Description of changes
* The `pcluster configure` command now includes prompts to enable EFA for supported instance-types
* These tests check to confirm if the wizard is able to generate the expected configuration file as well as successfully create a cluster with the same configuration

### Tests
* User is prompted to enable EFA for supported instance-type
* User is prompted to either provide an existing PlacementGroup or let PCluster create one for them
* A configuration file with expected properties is generated
* The configuration file can be used to successfully create a cluster

Result | Test | Duration | 
-- | -- | -- |
Passed | configure/test_pcluster_configure.py::test_efa_and_placement_group[us-east-1-c5n.9xlarge-alinux2-slurm-n-efa_config2-none] | 464.64 |  
Passed | configure/test_pcluster_configure.py::test_efa_and_placement_group[us-east-1-c5n.9xlarge-alinux2-slurm-y-efa_config0-default] | 492.55 |  
Passed | configure/test_pcluster_configure.py::test_efa_and_placement_group[us-east-1-c5n.9xlarge-alinux2-slurm-y-efa_config1-custom] | 466.77



### References
* [Prompt user to enable EFA for supported instance-types when using the pcluster configure wizard](https://github.com/aws/aws-parallelcluster/pull/4118)
* [ParallelCluster EFA](https://docs.aws.amazon.com/parallelcluster/latest/ug/efa.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

gmarciani commented on tests/integration-tests/tests/configure/test_pcluster_configure.py:
```diff
@@ -19,23 +19,48 @@
+        "skip_for_batch": True,
+    },
+    "enable_efa": lambda response: {
+        "prompt": r"Enable EFA on compute instances \(EFA\) \(y/n\) \[y\]:",
```
Why do we need the ending `(EFA)` in the prompt? It sounds redundant.

EddyMM replied:
This will be re-worded. For now I'll update it to use a regex that matches both the current and updated wording.

gmarciani commented on tests/integration-tests/tests/configure/test_pcluster_configure.py:
```diff
@@ -96,27 +117,149 @@ def test_region_without_t2micro(vpc_stack, pcluster_config_reader, key_name, reg
+
+
+@pytest.mark.parametrize(
+    "efa_response, efa_config, placement_group_response_type",
```
What do you mean with `*_response`?

EddyMM replied:
The response to the EFA prompt. Using something like `efa_enabled` would be confusing since one would expect a boolean. In this case however we're anticipating a "y" or "n". Hence the choice to use `efa_response`

gmarciani replied:
Got it: you mean the end.user response to the configure prompt

gmarciani commented on tests/integration-tests/conftest.py:
```diff
@@ -1246,6 +1247,25 @@ def network_interfaces_count(request, instance, region):
 
 
+@pytest.fixture(scope="class")
+def placement_group_stack(cfn_stacks_factory, request, region):
```
What is the benefit of moving this function from `tests/integration-tests/tests/networking/test_placement_group.py`?

EddyMM replied:
To make the stack available for other tests.
In this case the `pcluster configure` test needs to use a placement group that already exists.

gmarciani commented on tests/integration-tests/configs/common/common.yaml:
```diff
@@ -110,6 +110,13 @@ configure:
+  test_pcluster_configure.py::test_efa_and_placement_group:
+    dimensions:
+      - regions: ["us-east-1"]
+        instances: {{ common.INSTANCES_EFA_SUPPORT_X86 }}
```
Shouldn't we test both the case of instances supporting EFA and instances not supporting it?

EddyMM replied:
All the other pcluster configure tests use instance types that don't support EFA so I figured we could save some time and resources by skipping this. I could add the test in order to be thorough though.

gmarciani replied:
Having other tests use non-EFA instances is enough as long as they verify that the user is not prompted with EFA related questions.

gmarciani commented on tests/integration-tests/tests/configure/test_pcluster_configure.py:
```diff
@@ -96,27 +117,149 @@ def test_region_without_t2micro(vpc_stack, pcluster_config_reader, key_name, reg
     config_path = test_datadir / "config.yaml"
-    stages = orchestrate_pcluster_configure_stages(
-        region,
+    region_without_t2micro_prompts = (
```
Wy do we need to test region with absence of t2.micro?

EddyMM replied:
This is from a test that was already part of the suite. I only refactored it.
The test is meant to confirm that the default instance-type is retrieved dynamically. E.g. For regions that don't have `t2.micro`, the default instance-type is `t3.micro`.

gmarciani replied:
Got it thanks.

gmarciani commented on tests/integration-tests/tests/configure/test_pcluster_configure.py:
```diff
@@ -19,23 +19,48 @@
 from assertpy import assert_that
 from conftest import inject_additional_config_settings
 
+prompts = {
```
Since this is a constant, what about naming with uppercase?

EddyMM committed:
[PCluster Configure] Add integration tests for enabling EFA through the pcluster configure wizard

The `pcluster configure` command now includes prompts to enable EFA for supported instance-types. These tests check to confirm if the wizard is able to generate the expected configuration file as well as successfully create a cluster with the same configuration.

EddyMM committed:
Merge branch 'develop' into wip/integ-tests-enable-efa
